### DISCUSSION TITLE
CONFIG_DYNAMIC_DEBUG for linux dyndbg "Dynamic Debug" feature

### DISCUSSION THREAD

**Discussion by jtmoon79** [MichaIng/DietPi#5847] [Q&A] [answered]
**tl;dr** How do I get a `CONFIG_DYNAMIC_DEBUG` Linux build (Dynamic Debug) using DietPI ?

I have two devices running DietPi, a Raspberry Pi 4 and a NanoPI R2S. Both devices run DietPi on Debian Bullseye and are fully updated. But only one device has Linux "Dynamic Debug" enabled. I'm describing both devices here in case their upgrade path or hardware is the cause of the difference.

<br/>

### Nano Pi running Bullseye with Dynamic Debug

I installed DietPi on the Nano PI R2S. I installed Debian Bullseye.

The Nano PI R2S has the Linux [Dynamic Debug](https://www.kernel.org/doc/html/v5.15/admin-guide/dynamic-debug-howto.html) enabled. I am able to turn on kernel log messages for modules by writing commands to file `/sys/kernel/debug/dynamic_debug/control`. For example, to enable `wireguard` module kernel debug messages, I run command

```text
echo module wireguard +p > /sys/kernel/debug/dynamic_debug/control
```

<br/>

### Raspberry Pi running upgraded Bullseye without Dynamic Debug (missing `/sys/kernel/debug/dynamic_debug/`)

I installed DietPi on the Raspberry Pi 4 some years ago. I installed Debian Buster. I have since upgraded to Bullseye.

The Raspberry Pi 4 does not have file `/sys/kernel/debug/dynamic_debug/control` nor directory `/sys/kernel/debug/dynamic_debug/` (it does have `debugfs` mounted at path `/sys/kernel/debug/` with various files and directories, so the `debugfs` is working as expected).

I attempted to enable the "Dynamic Debug" feature on the Raspberry Pi 4. To the file `/boot/cmdline.txt`, I added [`wireguard.dyndbg=+p`](https://www.kernel.org/doc/html/v5.15/admin-guide/kernel-parameters.html) (to allow Dynamic Debug messages for the `wireguard` kernel module).
The `dmesg` logs shows my added command-line parameter is read by the kernel (so my edit of `cmdline.txt` was okay).

```text
[    0.000000] Kernel command line: … wireguard.dyndbg=+p
```

But a later `dmesg` log message reads
```text
[   12.307901] dyndbg param is supported only in CONFIG_DYNAMIC_DEBUG builds
```

Do you know how I enable "Dynamic Debug" for DietPi, i.e. how do I get a Linux kernel built with `CONFIG_DYNAMIC_DEBUG` ?

**Comment by Joulinar** [comment, marked as answer]
This is as expected because the native RPi kernel did not contain this feature. You would need to build your own RPI kernel or ask RPi developer if they could include this in a future release. 

It has been asked a year ago already, but got closed https://github.com/raspberrypi/linux/issues/3486

**Reply by jtmoon79** [reply]
Thanks for the link @Joulinar .

I'm wondering, how did I manage to install DietPi with a Linux kernel with Dynamic Debug enabled?

Looking at `/boot/`, I found on the Nano Pi
```text
$ grep -Fie 'DYNAMIC_DEBUG' -R /boot
/boot/config-5.15.75-rockchip64:CONFIG_DYNAMIC_DEBUG=y
/boot/config-5.15.75-rockchip64:CONFIG_DYNAMIC_DEBUG_CORE=y
```
I do not recall adding such a setting during DietPi installation (does DietPi offer this?).

...
_time passes_
...

Looking closer, the difference was staring at me for a long time.

The Nano Pi OS is `Debian`
```text
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
```
whereas the Raspberry Pi OS is `Raspbian`
```text
$ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
```

My eyes just glanced over this about one hundred times. 🤦

**Reply by Joulinar** [reply]
Quite simple, a R2S and a RPi4 are two complete different SBC using 2 different images and different kernel. We as DietPi team don't do any kernel development. We always use the native kernel for the underlying base image. On R2S, it should be the Armenian kernel if I'm not mistaken. While on Raspberry Pi, it's the own RPi kernel. As they are developed by different groups, it could easily happen that different parameters used.
